src/model_selection/crossValScore.ts

```diff
@@ -0,0 +1,219 @@
+ *                 the {@link CrossValidator} `cv`
+ */
+export async function crossValScore(
+  estimator: {
+    fit(X: Tensor2D, y: Tensor1D): Promise<unknown>
+    score(X: Tensor2D, y: Tensor1D): Scalar
+  },
+  Xy: [Scikit2D, Scikit1D],
+  params: {
+    cv?: CrossValidator
+    groups?: Scikit1D
+  }
+): Promise<Tensor1D>
+
+/**
+ * Evaluates a score by cross-validation. This particular overload
+ * of the function uses the default score of an unsupervised estimator
+ * for scoring.
+ *
+ * @param estimator A supervised estimator that has an async `fit(X)`
+ *                 function and a `score(X)` function.
+ *
+ * @param X        An single element array containing the features of the
+ *                 cross validation dataset.
+ *
+ * @param params   Cross validation parameters. `cv` is the {@link CrossValidator}
+ *                 responsible for splitting the dataset into training in test
+ *                 data in one or more folds. `groups` is additional grouping
+ *                 information which is used by some {@link CrossValidator}
+ *                 instance to determine which groups of datapoints may not
+ *                 appear both in test and training data at the same time.
+ *
+ * @returns        The test score for each split/fold that was generated by
+ *                 the {@link CrossValidator} `cv`
+ */
+export async function crossValScore(
+  estimator: {
+    fit(X: Tensor2D): Promise<unknown>
+    score(X: Tensor2D): Scalar
+  },
+  X: [Scikit2D],
+  params: {
+    cv?: CrossValidator
+    groups?: Scikit1D
+  }
+): Promise<Tensor1D>
+
+export async function crossValScore(
+  estimator: any,
+  [X, y]: [Scikit2D, Scikit1D?],
+  {
+    cv = new KFold(),
+    groups,
+    scoring
+  }: {
+    cv?: CrossValidator
+    groups?: Scikit1D
+    scoring?: any
+  } = {}
+): Promise<Tensor1D> {
+  if (undefined === scoring) {
+    assert(
+      'function' === typeof estimator.score,
+      'crossValScore(estimator,[X,y],params): Either params.scoring or estimator.score(X,y) must be defined.'
+    )
+    scoring = estimator.score
+  }
+
+  const scores: Scalar[] = []
+  scoring = scoring.bind(estimator)
+
+  let result: Tensor1D | undefined = undefined
+
+  tf.engine().startScope()
+  try {
+    X = convertToTensor2D(X)
+    if (null != y) {
+      y = convertToTensor1D(y)
+    }
+
+    for (const { trainIndex, testIndex } of cv.split(X, y, groups)) {
+      let score: Scalar | undefined
+
+      tf.engine().startScope()
+      try {
+        await estimator.fit(X.gather(trainIndex), y?.gather(trainIndex))
+
+        score = scoring(X.gather(testIndex), y?.gather(testIndex)) as Scalar
+        scores.push(score)
+      } finally {
+        tf.engine().endScope(score)
+        trainIndex.dispose()
+        testIndex.dispose()
+      }
+    }
+
+    return (result = tf.stack(scores) as Tensor1D)
+  } finally {
+    tf.engine().endScope(result)
```

Comment:
Same for the outer `try`. I'm just basically calling dispose on the Tensors to avoid the `try / finally / tf.engine` block. Here's the full code.

```
  X = convertToTensor2D(X)
  if (y != null) {
    y = convertToTensor1D(y)
  }

  for (const { trainIndex, testIndex } of cv.split(X, y, groups)) {
    let XTrain = X.gather(trainIndex)
    let yTrain = y?.gather(trainIndex)
    let XTest = X.gather(testIndex)
    let yTest = y?.gather(testIndex)

    await estimator.fit(XTrain, yTrain)
    let score = scoring(XTest, yTest) as Scalar

    scores.push(score)

    let cleanup = [XTrain, yTrain, XTest, yTest, trainIndex, testIndex]
    cleanup.forEach((el) => el?.dispose())
  }

  let cleanup = [X, y]
  cleanup.forEach((el) => el?.dispose())
  return tf.stack(scores) as Tensor1D
}
```

Reply:
Also you'll notice that I just create this intermediate arrays and loop through them just optionally calling dispose. We could just create a helper function in `utils` called `cleanup` that does just that, and use it here, which would save a couple more lines.

Basically,
```
function cleanup(tensors: any[]){
  tensors.forEach(el => el?.dispose())
}
```

And then the above would just be
`cleanup([X, y])` or `cleanup([XTrain, yTrain, blah, blah])`

Reply:
> There might be a way to get rid of the outer `try` / `finally` / `tf.engine`, by just manually calling dispose. I personally think it is easier to reason about, then nesting a double try here.
> 
> ```
>   X = convertToTensor2D(X)
>   if (y != null) {
>     y = convertToTensor1D(y)
>   }
> 
>   for (const { trainIndex, testIndex } of cv.split(X, y, groups)) {
>     let XTrain = X.gather(trainIndex)
>     let yTrain = y?.gather(trainIndex)
>     let XTest = X.gather(testIndex)
>     let yTest = y?.gather(testIndex)
> 
>     await estimator.fit(XTrain, yTrain)
>     let score = scoring(XTest, yTest) as Scalar
> 
>     scores.push(score)
> 
>     let cleanup = [XTrain, yTrain, XTest, yTest, trainIndex, testIndex]
>     cleanup.forEach((el) => el?.dispose())
>   }
> 
>   let cleanup = [X, y]
>   cleanup.forEach((el) => el?.dispose())
>   return tf.stack(scores) as Tensor1D
> }
> ```

@dcrescim I just want to point out the use of `tf.engine()` ; I was doing a project where I was having a memory leak. i tried out `tf.tidy` and `dispose` and yet there was still a bit of memory leak. but with `tf.engine()` there was no memory leak.

Reply:
`try {} finally {}` is a language feature and therefore not really replaceable by anything. You could hide it behind some `tidyAsync`. But that would likely just be more verbose. Even if we switch from `beginScope()` and `endScope()` to dispose, `try {} finally {}` remains irreplaceable. Just look at the following code:

```js
X = convertToTensor2D(X)
y = convertToTensor1D(y)
X.dispose()
y.dispose()
```
If the second line `y = convertToTensor1D(y)` throws an `IllegalArgumentException` because the input is for some reason invalid, then `X` is never disposed, polluting memory for the rest of the runtime.

TFJS' explicit memory management is a nightmare, especially since JS in is not at all suited for anything like this. Using `beginScope()` and `endScope()` do bring the risk other async() code of some other function being executed in the middle of the scope of this function. Basically all calls would have to go through a `mutex`, which might be implementable in terms of a `tidyAsync` function. But it's still kind of nightmarish.

Calling `dispose()` explicitly, however, is also kind of bad:
  * Prevents the usage of `tf.keep`
  * Will not clean up any memory allocated by `estimator.fit`
  * Makes it very easy to forget one tensor or another

Reply:
I've gotten rid of one try {} finally {} at least... It's still not ideal but I hope it will do until we have a more standard concept of how we want to handle memory.

Reply:
Works for me